File: app/src/main/java/com/botpa/turbophotos/sync/SyncActivity.java
package com.botpa.turbophotos.sync;

import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.botpa.turbophotos.R;
import com.botpa.turbophotos.sync.logs.LogAdapter;
import com.botpa.turbophotos.sync.users.User;
import com.botpa.turbophotos.sync.users.UserAdapter;
import com.botpa.turbophotos.util.Orion;
import com.botpa.turbophotos.util.Storage;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("SameParameterValue")
public class SyncActivity extends AppCompatActivity {

    //Connection
    private static final int STATUS_OFFLINE = 0;
    private static final int STATUS_CONNECTING = 1;
    private static final int STATUS_ONLINE = 2;

    private static final String CODE_CHARSET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    private int connectionStatus = STATUS_OFFLINE;

    //Users
    private final List<User> users = new ArrayList<>();
    private UserAdapter usersAdapter;

    //Logs
    private static final int logsMax = 200;
    private final List<String> logs = new ArrayList<>();
    private LogAdapter logsAdapter;

    //Views (connect)
    private View usersLayout;
    private EditText usersName;
    private EditText usersCode;
    private View usersConnect;
    private View usersLoading;
    private RecyclerView usersList;

    //Views (logs)
    private RecyclerView logsList;
    private View logsExit;


    //State
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.sync_screen);

        //Get views & add listeners
        getViews();
        addListeners();

        //Load users
        loadUsers();

        //Init users list
        initUsersList();

        //Init logs list
        initLogsList();

        //Init events observer
        initEventsObserver();

        //Start service
        Intent intent = new Intent(this, SyncService.class);
        startService(intent);
    }

    @Override
    protected void onDestroy() {
        //Close service
        send("stop");

        super.onDestroy();
    }

    //App
    private void getViews() {
        //Views (connect)
        usersLayout = findViewById(R.id.usersLayout);
        usersName = findViewById(R.id.usersName);
        usersCode = findViewById(R.id.usersCode);
        usersConnect = findViewById(R.id.usersConnect);
        usersLoading = findViewById(R.id.usersLoading);
        usersList = findViewById(R.id.usersList);

        //Views (logs)
        logsList = findViewById(R.id.logsList);
        logsExit = findViewById(R.id.logsExit);

        //Insets
        Orion.addInsetsChangedListener(getWindow().getDecorView(), new int[] { WindowInsetsCompat.Type.systemBars(), WindowInsetsCompat.Type.ime() });
    }

    private void addListeners() {
        //Connect
        usersCode.setOnKeyListener((view, i, keyEvent) -> {
            if (keyEvent.getKeyCode() == KeyEvent.KEYCODE_ENTER) {
                Orion.clearFocus(SyncActivity.this);
                Orion.hideKeyboard(SyncActivity.this);
                usersConnect.performClick();
            }
            return false;
        });

        usersConnect.setOnClickListener(view -> {
            //Already trying to connect
            if (connectionStatus != STATUS_OFFLINE) return;

            //Get code
            String code = usersCode.getText().toString();
            if (code.isEmpty()) return;

            //Get name
            String name = usersName.getText().toString();
            if (!name.isEmpty()) {
                //Check if name is saved
                boolean isSaved = false;
                for (int i = 0; i < users.size(); i++) {
                    //Get user
                    User user = users.get(i);
                    if (!user.name.equals(name)) continue;

                    //Update user
                    user.code = code;
                    usersAdapter.notifyItemChanged(i);
                    saveUsers();
                    isSaved = true;
                    break;
                }

                //Add new user
                if (!isSaved) {
                    users.add(0, new User(name, code));
                    usersAdapter.notifyItemInserted(0);
                    saveUsers();
                }
            }

            //Connect
            connect(code);
        });

        //Logs
        logsExit.setOnClickListener(view -> finish());
    }

    //Users
    private void initUsersList() {
        //Create adapter
        usersAdapter = new UserAdapter(SyncActivity.this, users);

        //Add connect listener
        usersAdapter.setOnClickListener((view, index) -> {
            //Get user
            User user = users.get(index);

            //Move user first in list
            users.remove(index);
            usersAdapter.notifyItemRemoved(index);
            users.add(0, user);
            usersAdapter.notifyItemInserted(0);
            saveUsers();

            //Connect to user
            connect(user.code);
        });

        //Add select user listener
        usersAdapter.setOnLongClickListener((view, index) -> {
            //Get user
            User user = users.get(index);

            //Select user info
            usersName.setText(user.name);
            usersCode.setText(user.code);

            //Hide keyboard
            Orion.hideKeyboard(SyncActivity.this);
            Orion.clearFocus(SyncActivity.this);
        });

        //Add delete user listener
        usersAdapter.setOnDeleteListener((view, index) -> {
            //Delete user
            users.remove(index);
            usersAdapter.notifyItemRemoved(index);
            saveUsers();
        });

        //Add adapter to list
        usersList.setAdapter(usersAdapter);
        usersList.setLayoutManager(new LinearLayoutManager(SyncActivity.this));
    }

    private void loadUsers() {
        //Var to check if save needed
        boolean needsSave = false;

        //Split users
        List<String> userStrings = Storage.getStringList("Sync.users");
        for (String userString: userStrings) {
            //Check if valid
            int separatorIndex = userString.indexOf("\n");
            if (separatorIndex == -1) {
                needsSave = true;
                continue;
            }

            //Add user
            users.add(new User(userString.substring(0, separatorIndex), userString.substring(separatorIndex + 1)));
        }

        //Save users after change
        if (needsSave) saveUsers();
    }

    private void saveUsers() {
        //Create string users
        List<String> userStrings = new ArrayList<>();
        for (User user: users) userStrings.add(user.toString());

        //Save list
        Storage.putStringList("Sync.users", userStrings);
    }

    //Connection
    private static long decodeBase36(String code) {
        long result = 0;
        for (int i = 0; i < code.length(); i++) {
            result = result * 36 + CODE_CHARSET.indexOf(code.charAt(i));
        }
        return result;
    }

    private String convertCodeToAddress(String code) {
        long combined = decodeBase36(code.toUpperCase());

        int port = (int) (combined & 0xFFFF);
        long ipNum = combined >> 16;

        String ip = String.format("%d.%d.%d.%d", (ipNum >> 24) & 0xFF, (ipNum >> 16) & 0xFF, (ipNum >> 8) & 0xFF, ipNum & 0xFF);

        return ip + ":" + port;
    }

    private void connect(String code) {
        //Hide keyboard
        Orion.hideKeyboard(SyncActivity.this);
        Orion.clearFocus(SyncActivity.this);

        //Connect
        if (code.contains(":")) {
            //Code is already an IP:PORT address
            send("connect", code);
        } else {
            //Code needs conversion to address
            send("connect", convertCodeToAddress(code));
        }
    }

    //Logs
    private void initLogsList() {
        //Create adapter
        logsAdapter = new LogAdapter(SyncActivity.this, logs);

        //Add adapter to list
        logsList.setAdapter(logsAdapter);
        logsList.setLayoutManager(new LinearLayoutManager(SyncActivity.this, LinearLayoutManager.VERTICAL, true));
        logsList.setItemAnimator(null);
    }

    private void log(String log) {
        //Reached maximum size -> Remove first
        if (logs.size() >= logsMax) {
            logs.remove(logs.size() - 1);
            logsAdapter.notifyItemRemoved(logs.size() - 1);
        }

        //Add new log
        logs.add(0, log);
        logsAdapter.notifyItemInserted(0);

        //Scroll to bottom
        logsList.scrollToPosition(0);
    }

    //Events
    private void initEventsObserver() {
        //Get instance
        SyncEventBus instance = SyncEventBus.getInstance();

        //Observe
        instance.trigger.observe(this, t -> {
            SyncEvent event;
            while ((event = instance.eventQueue.poll()) != null) handleEvent(event);
        });
    }

    private void handleEvent(SyncEvent event) {
        if (event == null) return;
        String command = event.command;

        //Check command
        switch (command) {
            //Service started
            case "init":
                log("Service started");
                break;

            //Status changed
            case "status":
                connectionStatus = event.valueInt;
                switch (connectionStatus) {
                    case STATUS_OFFLINE:
                        log("Disconnected");
                        usersLayout.setVisibility(View.VISIBLE);
                        usersLoading.setVisibility(View.GONE);
                        break;
                    case STATUS_CONNECTING:
                        log("Connecting...");
                        usersLoading.setVisibility(View.VISIBLE);
                        break;
                    case STATUS_ONLINE:
                        log("Connected");
                        usersLayout.setVisibility(View.GONE);
                        usersLoading.setVisibility(View.GONE);
                        break;
                }
                break;

            //Snack
            case "snack":
                Orion.snack(SyncActivity.this, event.valueString);
                break;

            //Log
            case "log":
                log(event.valueString);
                break;
        }
    }

    private void send(String command) {
        Intent intent = new Intent(this, SyncService.class);
        intent.putExtra("command", command);
        startService(intent);
    }

    private void send(String command, String value) {
        Intent intent = new Intent(this, SyncService.class);
        intent.putExtra("command", command);
        intent.putExtra("value", value);
        startService(intent);
    }

}

File: app/src/main/java/com/botpa/turbophotos/sync/SyncActivity.kt
package com.botpa.turbophotos.sync

import android.app.Activity
import android.content.Intent
import android.os.Bundle

import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.WindowInsetsSides
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.only
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawing
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.Observer

import com.botpa.turbophotos.gallery.views.Group
import com.botpa.turbophotos.gallery.views.GroupDivider
import com.botpa.turbophotos.gallery.views.GroupItems
import com.botpa.turbophotos.gallery.views.GroupTitle
import com.botpa.turbophotos.gallery.views.Layout
import com.botpa.turbophotos.gallery.views.groupItemPaddingHorizontal
import com.botpa.turbophotos.gallery.views.groupItemPaddingVertical
import com.botpa.turbophotos.settings.SettingsPairs
import com.botpa.turbophotos.sync.SyncEventBus.Companion.instance
import com.botpa.turbophotos.theme.CoonTheme
import com.botpa.turbophotos.theme.FONT_COMFORTAA
import com.botpa.turbophotos.theme.FONT_POPPINS
import com.botpa.turbophotos.util.Orion
import com.botpa.turbophotos.util.Storage

import java.util.Locale

class SyncActivity : ComponentActivity() {

    //View model
    private val view: SyncViewModel by viewModels()

    //Logs
    private val logsMax = 500


    //App
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        //Init storage
        Storage.init(this@SyncActivity)

        //Edging
        enableEdgeToEdge()

        //Load users
        loadUsers()

        //Init components
        initEventsObserver()

        //Content
        setContent {
            CoonTheme {
                SettingsLayout()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        //Close service
        send("stop")
    }

    //Layout
    @Composable
    private fun SettingsLayout() {
        Layout("Sync") {
            when (view.connectionStatus) {
                SyncService.STATUS_OFFLINE -> ConnectLayout(it, this, false)
                SyncService.STATUS_CONNECTING -> ConnectLayout(it, this, true)
                SyncService.STATUS_ONLINE -> LogsLayout(it)
            }
        }
    }

    @Composable
    private fun ConnectLayout(it: PaddingValues, activity: Activity, connecting: Boolean) {
        //Layout
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(it)
                .padding(horizontal = 20.dp)
                .windowInsetsPadding(WindowInsets.safeDrawing.only(WindowInsetsSides.Bottom))
        ) {
            //Connect
            Group {
                //Title
                GroupTitle("Connect")

                //Items
                GroupItems {
                    //Name input
                    TextField(
                        value = view.connectName,
                        label = { Text("Name (optional)") },
                        maxLines = 1,
                        onValueChange = { newValue: String -> view.connectName = newValue },
                        colors = TextFieldDefaults.colors(
                            focusedContainerColor = Color.Transparent,
                            unfocusedContainerColor = Color.Transparent,
                            disabledContainerColor = Color.Transparent,
                            focusedIndicatorColor = Color.Transparent,
                            unfocusedIndicatorColor = Color.Transparent
                        ),
                        modifier = Modifier
                            .fillMaxWidth()
                    )

                    //Divider
                    GroupDivider()

                    //Code input
                    TextField(
                        value = view.connectCode,
                        label = { Text("Code") },
                        maxLines = 1,
                        onValueChange = { newValue: String -> view.connectCode = newValue },
                        colors = TextFieldDefaults.colors(
                            focusedContainerColor = Color.Transparent,
                            unfocusedContainerColor = Color.Transparent,
                            disabledContainerColor = Color.Transparent,
                            focusedIndicatorColor = Color.Transparent,
                            unfocusedIndicatorColor = Color.Transparent
                        ),
                        modifier = Modifier
                            .fillMaxWidth()
                    )
                }

                //Connect button & connecting indicator
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(IntrinsicSize.Min)
                        .padding(top = 10.dp)
                ) {
                    //Connect button
                    Button(
                        onClick = {
                            //Already trying to connect
                            if (view.connectionStatus != SyncService.STATUS_OFFLINE) return@Button

                            //Check if code is valid
                            if (view.connectCode.isEmpty()) return@Button

                            //Check if name should get saved
                            if (!view.connectName.isEmpty()) {
                                //Check if name is saved
                                var isSaved = false
                                for (i in view.users.indices) {
                                    //Get user
                                    val user = view.users[i]
                                    if (user.name != view.connectName) continue

                                    //Update user
                                    user.code = view.connectCode
                                    saveUsers()
                                    isSaved = true
                                    break
                                }

                                //Add new user
                                if (!isSaved) {
                                    view.users.add(0, User(view.connectName, view.connectCode))
                                    saveUsers()
                                }
                            }

                            //Connect
                            connect(view.connectCode)
                        },
                        modifier = Modifier
                            .weight(1.0f)
                    ) {
                        Text(
                            text = "Connect",
                            fontFamily = FONT_COMFORTAA,
                            fontSize = 14.sp
                        )
                    }

                    //Connecting indicator
                    if (connecting) {
                        CircularProgressIndicator(
                            color = MaterialTheme.colorScheme.secondary,
                            trackColor = MaterialTheme.colorScheme.surfaceVariant,
                            strokeWidth = 4.dp,
                            modifier = Modifier
                                .padding(start = 10.dp)
                                .size(40.dp)
                        )
                    }
                }
            }

            //Users
            LazyColumn(
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f)
            ) {
                item {
                    Group {
                        //Title
                        GroupTitle("Users")

                        //Items
                        GroupItems {
                            if (view.users.isEmpty()) {
                                //Empty users message
                                Text(
                                    text = "There are no saved users",
                                    modifier = Modifier
                                        .padding(horizontal = groupItemPaddingHorizontal, vertical = groupItemPaddingVertical)
                                )
                            } else {
                                view.users.forEachIndexed { index, user ->
                                    //Add item
                                    UserItem(
                                        index = index,
                                        user = user,
                                        onConnect = { index, user ->
                                            //Move user first in list
                                            view.users.removeAt(index)
                                            view.users.add(0, user)
                                            saveUsers()

                                            //Connect to user
                                            connect(user.code)
                                        },
                                        onSelect = { index, user ->
                                            //Select user info
                                            view.connectName = user.name
                                            view.connectCode = user.code

                                            //Hide keyboard
                                            Orion.hideKeyboard(activity)
                                            Orion.clearFocus(activity)
                                        },
                                        onDelete = { index ->
                                            //Delete user
                                            view.users.removeAt(index)
                                            saveUsers()
                                        }
                                    )

                                    //Add divider between items
                                    if (index < view.users.size - 1) GroupDivider()
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    @Composable
    private fun LogsLayout(it: PaddingValues) {
        //List stuff
        val listState = rememberLazyListState()

        LaunchedEffect(Unit) {
            view.scrollRequest.collect {
                listState.animateScrollToItem(0)
            }
        }

        //Layout
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(it)
                .padding(horizontal = 20.dp)
                .windowInsetsPadding(WindowInsets.safeDrawing.only(WindowInsetsSides.Bottom))
        ) {
            //Logs
            LazyColumn(
                state = listState,
                reverseLayout = true,
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f)
            ) {
                itemsIndexed(view.logs) { index, log ->
                    //Add log
                    Text(
                        text = log,
                        fontFamily = FONT_POPPINS,
                        fontSize = 14.sp,
                        modifier = Modifier
                            .padding(vertical = 5.dp)
                    )
                }
            }

            //Exit button
            Button(
                onClick = { finish() },
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 10.dp)
            ) {
                Text(
                    text = "Exit",
                    fontFamily = FONT_COMFORTAA,
                    fontSize = 14.sp
                )
            }
        }
    }

    //Connect & Users
    private fun loadUsers() {
        //Var to check if save needed
        var needsSave = false

        //Split users
        val userStrings: MutableList<String> = Storage.getStringList(SettingsPairs.SYNC_USERS_KEY)
        for (userString in userStrings) {
            //Check if valid
            val separatorIndex = userString.indexOf("\n")
            if (separatorIndex == -1) {
                needsSave = true
                continue
            }

            //Add user
            view.users.add(
                User(
                    userString.substring(0, separatorIndex),
                    userString.substring(separatorIndex + 1)
                )
            )
        }

        //Save users after change
        if (needsSave) saveUsers()
    }

    private fun saveUsers() {
        //Create string users
        val userStrings: MutableList<String> = ArrayList()
        for (user in view.users) userStrings.add(user.toString())

        //Save list
        Storage.putStringList(SettingsPairs.SYNC_USERS_KEY, userStrings)
    }

    private fun decodeBase36(code: String): Long {
        val codeCharset = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        var result: Long = 0
        for (i in 0..<code.length) result = result * 36 + codeCharset.indexOf(code[i])
        return result
    }

    private fun convertCodeToAddress(code: String): String {
        val combined: Long = decodeBase36(code.uppercase(Locale.getDefault()))

        val port = (combined and 0xFFFFL).toInt()
        val ipNum = combined shr 16

        val ip = String.format(
            Locale.US,
            "%d.%d.%d.%d",
            (ipNum shr 24) and 0xFFL,
            (ipNum shr 16) and 0xFFL,
            (ipNum shr 8) and 0xFFL,
            ipNum and 0xFFL
        )

        return "$ip:$port"
    }

    private fun connect(code: String) {
        //Hide keyboard
        Orion.hideKeyboard(this@SyncActivity)
        Orion.clearFocus(this@SyncActivity)

        //Connect
        if (code.contains(":")) {
            //Code is already an IP:PORT address
            send("connect", code)
        } else {
            //Code needs conversion to address
            send("connect", convertCodeToAddress(code))
        }
    }

    //Logs
    private fun log(log: String) {
        //Reached maximum size -> Remove first
        if (view.logs.size >= logsMax) view.logs.removeAt(view.logs.size - 1)

        //Add new log
        view.logs.add(0, log)

        //Scroll to bottom
        view.requestScrollToBottom()
    }

    //Events
    private fun initEventsObserver() {
        //Get instance
        val instance = instance

        //Observe
        instance.trigger.observe(this, Observer { t: Boolean ->
            var event: SyncEvent?
            while ((instance.eventQueue.poll().also { event = it }) != null) handleEvent(event)
        })
    }

    private fun handleEvent(event: SyncEvent?) {
        //Get command
        val command = event?.command ?: return

        //Check command
        when (command) {
            "init" -> log("Service started")
            "status" -> {
                view.connectionStatus = event.valueInt
                when (view.connectionStatus) {
                    SyncService.STATUS_OFFLINE -> log("Disconnected")
                    SyncService.STATUS_CONNECTING -> log("Connecting...")
                    SyncService.STATUS_ONLINE -> log("Connected")
                }
            }
            "snack" -> Orion.snack(this@SyncActivity, event.valueString)
            "log" -> log(event.valueString)
        }
    }

    private fun send(command: String) {
        val intent = Intent(this, SyncService::class.java)
        intent.putExtra("command", command)
        startService(intent)
    }

    private fun send(command: String, value: String) {
        val intent = Intent(this, SyncService::class.java)
        intent.putExtra("command", command)
        intent.putExtra("value", value)
        startService(intent)
    }

}